### PR TITLE
Add -s USE_COCOS2D (emscripten-ports)

### PR DESCRIPTION
**Pull request by caiiiycuk**
Hi! This is first try to add -s USE_COCOS2D. It's possible to add --js-library && --std=c++11 in automatic way?

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -731,6 +731,7 @@ var USE_BULLET = 0; // 1 = use bullet from emscripten-ports
 var USE_VORBIS = 0; // 1 = use vorbis from emscripten-ports
 var USE_OGG = 0; // 1 = use ogg from emscripten-ports
 var USE_FREETYPE = 0; // 1 = use freetype from emscripten-ports
+var USE_COCOS2D = 0; // 3 = use cocos2d v3 from emscripten-ports
```
Comment (kripken) [inline src/settings.js]: Why is this `3`? Is that the version number?

Comment (caiiiycuk) [inline src/settings.js]: Yes it is a version number (version of cocos2d)

File: tools/ports/cocos2d.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import os
+import shutil
+import logging
+import re
+
+TAG = '-master'
+
+def get(ports, settings, shared):
```
Comment (kripken) [inline tools/ports/cocos2d.py]: should be just 2 spaces and not 4

File: tools/ports/cocos2d.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import os
+import shutil
+import logging
+import re
+
+TAG = '-master'
```
Comment (kripken) [inline tools/ports/cocos2d.py]: in the other ports we don't include the `-` in TAG, for consistency I think it's best to do the same here

**Comment by kripken** [comment]
Nice!

For `std=c++11`, do you mean for building the library, or the user's code? For the library, can add it in `create()`, and for the users code, I don't think that should be needed, unless I am missing something?

For a js-library, I am not sure, but I think in `process_args` you can add arbitrary commandline params. However, it's possible that due to the order it might not work, and we need to think of something else. Is a js-library needed here, though - perhaps EM_ASM etc stuff is good enough?

**Comment by caiiiycuk** [comment]
std=c++11 - actually I realize that it not needed for user code and current solution is ok.

js-library:
* process-args want work, as I understand through this args I can add arg to clang compiler not to emcc.
* I should about EM_ASM. cocos2d required for some code to be implemented on client side (https://github.com/emscripten-ports/Cocos2d/blob/master/cocos2dx/platform/emscripten/CCTextureCacheEmscripten.js). May be I can eval this code in EM_ASM, but I should find good place to call EM_ASM. I mean that this code should be executed before any GL code.

**Comment by kripken** [comment]
If there is a constructor or init function, that might be an OK place to put it? If not, we probably need to add a `system_libs.get_js_libraries` method that `emcc.py` uses to query ports for extra js libs to add.

**Comment by caiiiycuk** [comment]
I fix formatting, however I am new in python, may be I do something wrong. Also I remove dependency from js-library. I am not sure that I do it right, because I call c++ function directly form EM_ASM, is it safe?
(https://github.com/emscripten-ports/Cocos2d/blob/master/cocos2dx/platform/emscripten/CCTextureCacheEmscripten.cpp#L222)

**Comment by caiiiycuk** [comment]
I know that I should add functions which I want call from js to EXPORTED_FUNCTIONS. Can I do this through modifying settings object. Or maybe I can pass function pointer to JS and call through that pointer?

**Comment by kripken** [comment]
`CCTextureCacheEmscripten_preMultiplyImageRegion` is a C++ method?  Then you should add `EMSCRIPTEN_KEEPALIVE` in the C++ before the function name, which keeps it alive and exports it. You can then call it from JS using `Module['_CCTextureCacheEmscripten_preMultiplyImageRegion']`.

**Comment by caiiiycuk** [comment]
Great! Fixed.

File: tools/ports/__init__.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import binaryen, bullet, freetype, libpng, ogg, sdl, sdl_image, sdl_ttf, sdl_net, vorbis, zlib
+import binaryen, bullet, freetype, libpng, ogg, sdl, sdl_image, sdl_ttf, sdl_net, vorbis, zlib, cocos2d
```
Comment (kripken) [inline tools/ports/__init__.py]: this list is sorted, please put it in the sorted location, after bullet.

File: tools/ports/cocos2d.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      o_s = []
+      for src in cocos2dx_src:
+        o = os.path.join(cocos2d_build, 'Cocos2d-' +
+                 TAG, 'build', src + '.o')
+        shared.safe_ensure_dirs(os.path.dirname(o))
+        command = [shared.PYTHON,
+               shared.EMCC,
+               os.path.join(cocos2dx_root, 'proj.emscripten', src),
+               '-Wno-overloaded-virtual',
+               '-U__native_client__',
```
Comment (kripken) [inline tools/ports/cocos2d.py]: why is this needed?

File: tools/ports/cocos2d.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+               '-Wno-deprecated-declarations',
+               '-DEMSCRIPTEN',
```
Comment (kripken) [inline tools/ports/cocos2d.py]: emcc will set this anyhow, so no need for it

**Comment by caiiiycuk** [comment]
Fixed

File: tools/ports/cocos2d.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      if add_next:
+        sources.append(
+          re.search('\s*(.*?)(\s*\\\\$|\s*$)', line, re.IGNORECASE).group(1))
+        add_next = line.endswith('\\\n')
+  
+  sources.append(os.path.join(cocos2d_root, 'CocosDenshion', 'emscripten', 'SimpleAudioEngine.cpp'))
+  return sources
+
+
+def make_incldues(cocos2d_root, cocos2dx_root):
```
Comment (kripken) [inline tools/ports/cocos2d.py]: should be `includes` (also where called)

**Comment by kripken** [comment]
Also please add a line to test_embuilder in test_sanity.py.

Aside from those last comments, looks great!

**Comment by caiiiycuk** [comment]
Fixed

**Comment by kripken** [comment]
Hmm, the test appears to show the right output, but fails on
````
--- expected
+++ actual
@@ -1 +1 @@
-/report_result?0
+/report_result?3
````
Perhaps we just need to add `reference_slack=3` to the `btest()` call, but maybe there's a way to make the output more identical across platforms?

**Comment by caiiiycuk** [comment]
In what environment test fails?

**Comment by caiiiycuk** [comment]
I should run pngcursh on result image. Can you attach image that produced when test fails?

**Comment by caiiiycuk** [comment]
Now PR contains image processed with pngcrush, hope that helps.

**Comment by kripken** [comment]
This is on linux with AMD graphics. Probably slight differences in the GPU output. Now I see just a difference of 1 versus 3 from before, so pngcrush has helped. (I wonder how? I'm not familiar with it.)

Anyhow this looks good to merge, please just add `reference_slack=1` to the `btest` call. (We need that in a few other GL tests too, the output is just not 100% deterministic across platforms.)

**Comment by caiiiycuk** [comment]
Fixed. I also does not know what pngcursh do. I just run it as said in tests/runner.py

**Comment by kripken** [comment]
Thanks. I'll look into pngcrush, now I'm curious...
